### ISSUE TITLE
Add Groovy DSL shugar to make kotlin.jvm.functions.Function[x] cleaner

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/gradle#7427]
## TL;DR

If I have an extension type `typeOf<() -> Unit>()` I add to the Project `ExtensionContainer`, in the Kotlin I can access nicely but in Groovy DSL I can not.
Kotlin DSL:
```kotlin
someAddedExtension()
```
Groovy DSL:
```groovy
someAddedExtension.invoke()
```

## Detail Dive

In kotlin these are the various functional interface types.
If you do `typeOf<() -> Unit>()` the type becomes `kotlin.jvm.functions.Function0<Unit>`

Thes are some of the functions and how they are implemented in Kotlin.
```kotlin
package kotlin.jvm.functions

/** A function that takes 0 arguments. */
public interface Function0<out R> : Function<R> {
    /** Invokes the function. */
    public operator fun invoke(): R
}
/** A function that takes 1 argument. */
public interface Function1<in P1, out R> : Function<R> {
    /** Invokes the function with the specified argument. */
    public operator fun invoke(p1: P1): R
}
/** A function that takes 2 arguments. */
public interface Function2<in P1, in P2, out R> : Function<R> {
    /** Invokes the function with the specified arguments. */
    public operator fun invoke(p1: P1, p2: P2): R
}
...
```

In groovy, if you have a function of type `Function0` you must invoke it using the `invoke` method.

I propose that Gradle offer some sugar on all of the Kotlin functional types to allow those functional types to be invoked like this:
```
someAddedExtension()
```

This sort of mechanism is already available if your extension is of type `java.util.concurrent.Callable<*>`. I'm not sure if this is a feature of Groovy or the Groovy Gradle DSL.

## Benefit

This will provide a mechanism for Kotlin DSL plugin authors to add functions to the `ExtensionContainer` and allow consumers to call them consistently in both Groovy and Kotlin.

**Comment by JLLeitschuh** [comment]
@oehme Do you know of any way that this could be done without adding a requirement in Gradle Core on the Kotlin standard library?

**Comment by big-guy** [comment]
@JLLeitschuh do you have an example of why you'd want to have a standalone method vs packaging this into an extension class?  You're going to be able to put javadoc on the extension class, call the method in a "normal" way from Groovy/Kotlin/Java and not worry about colliding with other methods that may be added later to Gradle's API.

e.g., instead of:
```
someAddedExtension()
```
You have:
```
extension {
   someAdded()
}
// or
extension.someAdded()
```

**Comment by JLLeitschuh** [comment]
## Origin
This whole thing spun off of this discussion:
https://gradle-community.slack.com/archives/CA7UM03V3/p1539784299000100

## Use Case
Yes, the use case is for this plugin:
https://github.com/wpilibsuite/GradleRIO/issues/136
(It's currently written in Groovy but there's a goal to convert it to the Kotlin DSL for next year).

That plugin adds an extension closure on the `DependencyHandler`.
This is where it is added:
https://github.com/wpilibsuite/GradleRIO/blob/03ce5bcddfb7d9fe98fcf9b17d2da44d1e0f353f/src/main/groovy/edu/wpi/first/gradlerio/wpi/dependencies/WPIJavaDeps.groovy#L43-L68

The API that these extensions enable in Groovy looks like this:
```groovy
dependencies {
    compile wpilib()
}
```

This is the same syntax that has been normalized by the `DependencyHandler` API.
Eg:
```
dependencies {
    compile gradleApi()
    testCompile gradleTestKit()
}
```

Plugins should be able to add these sort of extension functions to `DependencyHandler` and the API should look the same in Kotlin or in Groovy.


## Current state for Kotlin based Gradle Plugins

The simplest way to expose these sorts of extension aware extension functions when writing a Kotlin Plugin is to do something like this:

```kotlin
val somethingUsefulImpl = fun() {
    println("Hello world project ${project.name}!")
}
project
    .extensions
    .add(
        typeOf<() -> Unit>(),
        "doSomethingUseful",
        somethingUsefulImpl
   )
```

If you do this in a plugin you end up with the following extension being generated by the Kotlin DSL:
```kotlin
/**
 * Retrieves the [doSomethingUseful][kotlin.jvm.functions.Function0<kotlin.Unit>] extension.
 */
val org.gradle.api.Project.`doSomethingUseful`: kotlin.jvm.functions.Function0<kotlin.Unit> get() =
    (this as org.gradle.api.plugins.ExtensionAware).extensions.getByName("doSomethingUseful") as kotlin.jvm.functions.Function0<kotlin.Unit>
```

This is awesome because then in the Kotlin DSL you can just do this:
```kotlin
somethingUseful()
```
But in the Groovy DSL (currently) you still have to do this:
```groovy
somethingUseful.invoke()
```

**Comment by oehme** [comment]
If we want this, it should be based on core JVM types like Callable. Making it look pretty is part of each DSL.

**Comment by JLLeitschuh** [comment]
@oehme The problem with that is that the Kotlin functions go up to 22 arguments:
This is an example of `Function17`
```kotlin
/** A function that takes 17 arguments. */
public interface Function17<in P1, in P2, in P3, in P4, in P5, in P6, in P7, in P8, in P9, in P10, in P11, in P12, in P13, in P14, in P15, in P16, in P17, out R> : Function<R> {
    /** Invokes the function with the specified arguments. */
    public operator fun invoke(p1: P1, p2: P2, p3: P3, p4: P4, p5: P5, p6: P6, p7: P7, p8: P8, p9: P9, p10: P10, p11: P11, p12: P12, p13: P13, p14: P14, p15: P15, p16: P16, p17: P17): R
}
```
There is no equivalent to `Function[4-22]` in the java standard library.

**Comment by oehme** [comment]
Let's worry about that when someone actually needs it. I wouldn't want 4-22 arg function calls in my build scripts. We generally do callbacks with builder style APIs

**Comment by big-guy** [comment]
Is there something in some of the new platform/alignment dependency management work that would be a better replacement for `wpilib()`?

The `localGroovy()`, `gradleApi()`, `gradleTestKit()` aren't a great pattern to emulate and I think using normal dependency notations would be better here.

**Comment by JLLeitschuh** [comment]
Currently, `Callable` is the best replacement for our use case, but there is no replacement if there is a need for 3 or more args.
For Java 8: 1 arg could be covered by `java.util.function.Function<T, R>`, 2 args could be covered by `java.util.function.BiFunction<T, U, R>`.
I'm not certain ow does the Groovy/Kotlin DSL handles these types.
